### ISSUE TITLE
Support other hosting solutions (i.e. write drivers for them), such as Dropbox, Mediafire, MEGA...

### ISSUE THREAD

**Issue by Paturages** [Paturages/chorus#8]

**Comment by sandsoles16** [comment]
I don't think this could be possible for Mediafire, but I can look into MEGA's API.

**Comment by Paturages** [comment]
@sandsoles16 Good luck, it's a rather tough one because they only officially provide a C++ low-level API. Though they also provide a [higher level CLI utility](https://github.com/meganz/megacmd), which I'd advise you to play around with if you can. If you're able to do the 2 items described in the OP, that'd be awesome!

**Comment by shelvacu** [comment]
Yeah for mega you'll probably want to use the `megacli` binary, and for the rest you can probably use [plowshare](https://github.com/mcrapet/plowshare).

**Comment by Paturages** [comment]
@shelvacu `plowshare` looks really interesting, thanks!

**Comment by Paturages** [comment]
Yo, so as you can probably tell, this is a pretty stale issue, but the subject of torrents as means to share charts is interesting to talk about, debate and document at least.

Theoretically, as long as there were enough seeders to keep it going, it would be quite viable: if one were savvy enough, I think it would even be promising as a means to provide incremental updates to existing charts, such as official game rip fixes.

In practice, the community isn't really likely to take on "traditional torrenting" as a whole: as an end user, it adds friction and another barrier of entry to the community and is less straightforward than just clicking and downloading stuff on a browser. Morever, the image of torrenting reinforces the, well, gray area between fair use and outright pirating so to speak: one would feel like they're getting into naughty/shady things that way.

I don't think it's entirely hopeless, but the user experience has to be improved beforehand. I think with technological advances such as the rise of IPFS or other P2P protocols etc, we might be able to come up with something in the long run, but it has to be community consensus.

On the other hand, we should probably also strive for a friendlier solution in regards to artists and bands: back in the day, [JamLegend](https://en.wikipedia.org/wiki/JamLegend) seemed to be a promising model for the indie scene at least and at some point even attracted some major bands like Dragonforce, Stratovarius, Rise Against among others. That was around 2008-2010: nowadays, we are right in the middle of the "streaming bubble" with platforms such as SoundCloud, Bandcamp, Spotify, etc: I feel like there's promise in also looking in a genuine way to solve the rhythm game custom scene.

To come back to the issue itself, right now the community has resolved itself to "standardize" itself on Google Drive, and Google Drive is in practice the only driver in use right now. Other platforms are a bit less user friendly, and less dev friendly to go and fetch some items from (even though Google Drive is kind of a hot mess in itself). Some even seem shady nowadays because of the rise of Discord scams and so on. So right now I don't see the community parting with the "Google Drive only" mindset for now: you'd have to build something with a fairly strong value proposition over Google Drive to convince the entire community, but I don't think it's impossible.

**Comment by Dogeek** [comment]
Well I know torrenting is strongly associated with piracy, but I do think it should be a viable solution going forward with clone hero.

The advantages of a P2P network for charts are multiple :
- No more "this drive has too many downloads" issue
- It's a cleaner implementation in my opinion, especially if clone hero integrates natively with it
- If CH integrates the torrent protocol for charts, we'd be able to synchronize libraries from multiple people when playing online (no more 4/5000 songs playable in a lobby :) )

Of course, the ideal solution in my opinion, is to have soundcloud/bandcamp/deezer/spotify integrations for the music, and just have to download the chart and ini files which are a few kB. It also gives a kick back to the artists if we use those APIs as they will get their cut from the streaming service every time a song is played. But I'm getting side tracked.

I think there is not a lot of friction to adding magnet links to each song. We could implement the driver in such a way that the server downloads the charts of GDrive, and then seeds the torrents for the first few downloads until the torrent reaches critical mass to sustain itself without the original seeder. I also think a lot of people would download charts that way, since they make it easy to download one by one, instead of dealing with GDrive's awful UX. From my point of view, once the end user has a torrent client (and let's be honest, most people do), I'd argue that it's even more straightforward to download than using google drive (especially with magnet links)

**Comment by Paturages** [comment]
There's no chance of the game itself integrating torrenting for custom content as that would be a bit aggravating legally: CH is already pretty careful on that regard and has to stay as a platform to play custom songs, and will always delegate custom song management to other platforms, modules or even mods.

I think it might be doable to orchestrate a way to torrent custom songs through a mod, but I also think that seeding should be an opt-in process rather than an opt-out one since not everyone is comfortable with something leveraging their internet connection for stuff (background crypto miners anyone; it might also kind of affect game performance at a high enough level). Another approach for a mod would be for it to only trigger upon online matching to solve the case of 4/5000 available songs (not all of them at once, maybe at song selection only).

Adding a magnet/.torrent alongside regular download links could be an option, but you'd need to set up a tracker for that, leverage an existing one or just rely on DHT.

An idea to rethink the whole scanning process (it does need refactoring, no hard feelings) would be to actually keep the downloads and also seed them. It might be a good general idea paradigm wise: the host that would be scanning and updating the chorus database would only need to:
* Download all folders, subfolders and nested files for the provided Google Drives (or other things), or check if they're up to date if it's not the first run
* Scan their contents: charts, metadata, presence of certain files (stems, background, video, ...)
* Persist the data in the central database

So the general idea to refactor this whole project (I don't really have a will to maintain it anymore) would be to divide it into:
- A module responsible for downloading from multiple sources (it'd basically be JDownloader, but better hopefully). That alone would also be useful for end users who need a script to get a lot of songs at once, although that's also solvable through torrent eventually. This would also need to unzip archives automatically (doable using 7-zip).
- A module responsible for scanning the contents of a local folder and detecting and compiling all CH-compatible folders into, for example, a SQLite db file, a JSON or even CH's own song cache format. That would also be useful for end users who want to manage their local library, and if the export to CH song cache format is implemented, it can even be an optimization for scanning huge libraries (50,000+ songs). I do recommend looking into implementing that part in a lower level language like Rust, or something that deals with a lot of file I/O well (perhaps Go for concurrency, although in any case disk I/O will most likely end up being the bottleneck).
- An interface to browse a local library through the export generated by the above. This would essentially be equivalent to chorus's frontend, but repurposing that as a local library manager could be desirable: I know a fair amount of people who have ungodly amounts of songs they don't even know the full contents of.

Yes I know I'm getting extremely sidetracked, but I do think that at least starting with building a "JDownloader alternative" to have a fairly reliable way to download everything needed (it should still be less than 1 TB I'm pretty sure) is a fair first step as far as thinking about community torrents (since we do need the data to seed it beforehand). People have been playing around with chart formats as well, so this might be a good time to think about an independent/community folder/chart parser thing. And as far as chorus itself goes, I honestly haven't had the heart to evolve it as it is, so might as well think of something new to succeed it.
